File: java/v2.8/src/main/java/com/google/api/services/samples/dfareporting/conversions/UpdateOfflineMobileConversion.java
/*
 * Copyright (c) 2017 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations under
 * the License.
 */

package com.google.api.services.samples.dfareporting.conversions;

import com.google.api.services.dfareporting.Dfareporting;
import com.google.api.services.dfareporting.model.Conversion;
import com.google.api.services.dfareporting.model.ConversionError;
import com.google.api.services.dfareporting.model.ConversionStatus;
import com.google.api.services.dfareporting.model.ConversionsBatchUpdateRequest;
import com.google.api.services.dfareporting.model.ConversionsBatchUpdateResponse;
import com.google.api.services.dfareporting.model.FloodlightActivity;
import com.google.api.services.samples.dfareporting.DfaReportingFactory;
import com.google.common.collect.ImmutableList;

/**
 * This example updates the quantity and value of a conversion attributed to a mobile device ID. To
 * create a conversion attributed to a mobile device ID, run InsertOfflineMobileConversion.java.
 */
public class UpdateOfflineMobileConversion {
  private static final String USER_PROFILE_ID = "INSERT_USER_PROFILE_ID_HERE";

  // Values that identify an existing conversion.
  private static final String CONVERSION_MOBILE_ID = "INSERT_CONVERSION_MOBILE_ID_HERE";
  private static final String FLOODLIGHT_ACTIVITY_ID = "INSERT_FLOODLIGHT_ACTIVITY_ID_HERE";
  private static final String ORDINAL = "INSERT_ORDINAL_VALUE_HERE";
  private static final String TIMESTAMP = "INSERT_TIMESTAMP_IN_MICROSECONDS_HERE";

  // Values to update for the specified conversion.
  private static final String NEW_QUANTITY = "INSERT_NEW_CONVERSION_QUANTITY_HERE";
  private static final String NEW_VALUE = "INSERT_NEW_CONVERSION_VALUE_HERE";

  public static void runExample(
      Dfareporting reporting, long profileId, long floodlightActivityId, String mobileDeviceId,
      String ordinal, long timestampMicros, long newQuantity, double newValue) throws Exception {
    // Find Floodlight configuration ID based on the provided activity ID.
    FloodlightActivity floodlightActivity = reporting.floodlightActivities()
        .get(profileId, floodlightActivityId).execute();
    long floodlightConfigurationId = floodlightActivity.getFloodlightConfigurationId();

    // Create a conversion object populated with values that identify the conversion to update.
    Conversion conversion = new Conversion();
    conversion.setMobileDeviceId(mobileDeviceId);
    conversion.setFloodlightActivityId(floodlightActivityId);
    conversion.setFloodlightConfigurationId(floodlightConfigurationId);
    conversion.setOrdinal(ordinal);
    conversion.setTimestampMicros(timestampMicros);

    // Set the fields to be updated. These fields are required; to preserve a value from the
    // existing conversion, it must be copied over manually.
    conversion.setQuantity(newQuantity);
    conversion.setValue(newValue);

    ConversionsBatchUpdateRequest request = new ConversionsBatchUpdateRequest();
    request.setConversions(ImmutableList.of(conversion));

    ConversionsBatchUpdateResponse response = reporting.conversions()
        .batchupdate(profileId, request).execute();

    if (!response.getHasFailures()) {
      System.out.printf("Successfully updated conversion for mobile device ID %s.%n",
          mobileDeviceId);
    } else {
      System.out.printf("Error(s) updating conversion for mobile device ID %s:%n",
          mobileDeviceId);

      // Retrieve the conversion status and report any errors found. If multiple conversions
      // were included in the original request, the response would contain a status for each.
      ConversionStatus status = response.getStatus().get(0);
      for (ConversionError error : status.getErrors()) {
        System.out.printf("\t[%s]: %s.%n", error.getCode(), error.getMessage());
      }
    }
  }

  public static void main(String[] args) throws Exception {
    Dfareporting reporting = DfaReportingFactory.getInstance();

    long floodlightActivityId = Long.parseLong(FLOODLIGHT_ACTIVITY_ID);
    long newQuantity = Long.parseLong(NEW_QUANTITY);
    long profileId = Long.parseLong(USER_PROFILE_ID);
    long timestampMicros = Long.parseLong(TIMESTAMP);

    double newValue = Double.parseDouble(NEW_VALUE);

    runExample(reporting, profileId, floodlightActivityId, CONVERSION_MOBILE_ID,
        ORDINAL, timestampMicros, newQuantity, newValue);
  }
}

File: php/v3.4/examples/Conversions/UpdateOfflineMobileConversion.php
<?php
/*
 * Copyright 2017 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

// Require the base class.
require_once dirname(__DIR__) . '/BaseExample.php';

/**
 * This example updates an offline conversion attributed to a mobile device ID.
 * To create a conversion attributed to a mobile device ID, run
 * InsertOfflineMobileConversion.php.
 */
class UpdateOfflineMobileConversion extends BaseExample
{
    /**
     * (non-PHPdoc)
     * @see BaseExample::getInputParameters()
     * @return array
     */
    protected function getInputParameters()
    {
        return [['name' => 'user_profile_id',
                 'display' => 'User Profile ID',
                 'required' => true],
                ['name' => 'mobile_device_id',
                 'display' => 'Mobile Device ID',
                 'required' => true],
                ['name' => 'floodlight_activity_id',
                 'display' => 'Floodlight Activity ID',
                 'required' => true],
                ['name' => 'ordinal',
                 'display' => 'Ordinal',
                 'required' => true],
                ['name' => 'timestamp',
                 'display' => 'Timestamp (microseconds)',
                 'required' => true],
                ['name' => 'new_quantity',
                 'display' => 'New Quantity',
                 'required' => true],
                ['name' => 'new_value',
                 'display' => 'New Value',
                 'required' => true]];
    }

    /**
     * (non-PHPdoc)
     * @see BaseExample::run()
     */
    public function run()
    {
        $values = $this->formValues;

        printf(
            '<h2>Updating offline conversion for mobile device ID "%s"</h2>',
            $values['mobile_device_id']
        );

        // Find Floodlight configuration ID based on provided activity ID.
        $activity = $this->service->floodlightActivities->get(
            $values['user_profile_id'],
            $values['floodlight_activity_id']
        );
        $floodlightConfigId = $activity->getFloodlightConfigurationId();

        // Create a conversion object with values that identify the conversion to
        // update.
        $conversion = new Google_Service_Dfareporting_Conversion();
        $conversion->setMobileDeviceId($values['mobile_device_id']);
        $conversion->setFloodlightActivityId($values['floodlight_activity_id']);
        $conversion->setFloodlightConfigurationId($floodlightConfigId);
        $conversion->setOrdinal($values['ordinal']);
        $conversion->setTimestampMicros($values['timestamp']);

        // Set the fields to be updated. These fields are required; to preserve a
        // value from the existing conversion, it must be copied over manually.
        $conversion->setQuantity($values['new_quantity']);
        $conversion->setValue($values['new_value']);

        $batch = new Google_Service_Dfareporting_ConversionsBatchUpdateRequest();
        $batch->setConversions([$conversion]);

        $result = $this->service->conversions->batchupdate(
            $values['user_profile_id'],
            $batch
        );

        if (!$result->getHasFailures()) {
            printf(
                'Successfully updated conversion for mobile device ID %s.',
                $values['mobile_device_id']
            );
        } else {
            printf(
                'Error(s) updating conversion for mobile device ID %s:<br><br>',
                $values['mobile_device_id']
            );

            $status = $result->getStatus()[0];
            foreach ($status->getErrors() as $error) {
                printf('[%s] %s<br>', $error->getCode(), $error->getMessage());
            }
        }
    }

    /**
     * (non-PHPdoc)
     * @see BaseExample::getName()
     * @return string
     */
    public function getName()
    {
        return 'Update Offline Mobile Conversion';
    }
}
